core/src/main/java/org/infinispan/persistence/spi/CacheWriter.java

```diff
@@ -37,13 +50,27 @@
 
    /**
     * Persist all provided entries to the store in a single batch update. If this is not supported by the
-    * underlying store, then entries are written to the store individually via {@link #write(MarshalledEntry)}.
+    * underlying store, then entries are written to the store individually via {@link #write(MarshallableEntry)}.
     *
     * @param entries an Iterable of MarshalledEntry to be written to the store.
     * @throws NullPointerException if entries is null.
+    * @deprecated since 10.0, use {@link #writeBatch(Publisher)} instead.
     */
+   @Deprecated
    default void writeBatch(Iterable<MarshalledEntry<? extends K, ? extends V>> entries) {
-      entries.forEach(this::write);
+      entries.forEach(e -> write((MarshallableEntry<? extends K, ? extends V>) e));
+   }
+
+   /**
+    * Persist all provided entries to the store in a single batch update. If this is not supported by the
+    * underlying store, then entries are written to the store individually via {@link #write(MarshallableEntry)}.
+    *
+    * @param publisher a {@link Publisher} of {@link MarshallableEntry} instances
+    * @throws NullPointerException if the publisher is null.
+    */
+   default void writeBatch(Publisher<MarshallableEntry<? extends K, ? extends V>> publisher) {
+      // TODO transform MarshallableEntry to MarshalledEntry and call writeBatch(Iterable)
+      Flowable.fromPublisher(publisher).blockingForEach(this::write);
```

Comment:
@wburns I wanted this to process the publisher, wrap the MarshallableEntries as a MarshalledEntry if required and then return a `blockingIterable` to pass to `writeBatch`, but I couldn't get generics to play nice. I tried many variations of below without success, any ideas?

```java
writeBatch(
            Flowable.fromPublisher(publisher)
                  .map(e -> MarshalledEntry.wrap(e))
                  .blockingIterable()
      );
```

Reply:
So I am not liking the `void` return type. I am thinking maybe we want this to return a `CompletionStage<Void>`. This will play much nicer in the long run once we have a non blocking store SPI as well.

Looking re: default method.

Reply:
The problem is the ? extends, you normally pass in a concrete type. I am always disappointed with nested wildcard generics.

What you have there is fine, you will just have to trick the generics by adding in an erased cast, sadly.

Reply:
This seems to work:

```java
   default <K1 extends K, V1 extends V> void writeBatch(Iterable<MarshalledEntry<K1, V1>> entries) {
      entries.forEach(this::write);
   }
```

Reply:
After discussing this with Will I suggest renaming `writeBatch` to `writeBulk` and removing this from the javadoc:

```
 * Persist all provided entries to the store in a single batch update. If this is not supported by the
 * underlying store, then entries are written to the store individually via {@link #write(MarshallableEntry)}.
```